File: tests/python/run_qwen3.py
#!/usr/bin/env python3
import argparse
import os
import sys
import time
import torch

sys.path.insert(0, os.path.dirname(__file__))
from helpers import GlmOps, get_model_path
from qwen3_model import Qwen3Model, _normalize_token_ids

from transformers import AutoTokenizer

QWEN3_REPO = "Qwen/Qwen3-0.6B"
EOS_TOKEN_IDS = {151645, 151643}


def chat_loop(model, tokenizer, max_new_tokens, no_think):
    messages = []
    enable_thinking = not no_think

    while True:
        try:
            user_input = input("\nYou: ").strip()
        except (EOFError, KeyboardInterrupt):
            print()
            break

        if not user_input:
            continue
        if user_input.lower() in {"quit", "exit", "/q"}:
            break
        if user_input.lower() == "/clear":
            messages = []
            print("Conversation cleared.")
            continue

        messages.append({"role": "user", "content": user_input})

        try:
            input_ids = tokenizer.apply_chat_template(
                messages, tokenize=True, add_generation_prompt=True,
                enable_thinking=enable_thinking,
            )
        except TypeError:
            input_ids = tokenizer.apply_chat_template(
                messages, tokenize=True, add_generation_prompt=True,
            )

        input_ids = _normalize_token_ids(input_ids, tokenizer)
        input_tensor = torch.tensor([input_ids], dtype=torch.int64)

        if len(input_ids) > model.max_seq_len:
            print(f"Prompt ({len(input_ids)} tokens) exceeds max_seq_len ({model.max_seq_len}). Truncating conversation.")
            messages.pop()
            continue

        print("\nAssistant: ", end="", flush=True)
        generated_ids = []
        start = time.time()
        token_count = 0

        for token_id in model.generate_tokens(
            input_tensor, max_new_tokens=max_new_tokens,
            eos_token_ids=EOS_TOKEN_IDS,
        ):
            generated_ids.append(token_id)
            token_count += 1
            chunk = tokenizer.decode([token_id], skip_special_tokens=False)
            print(chunk, end="", flush=True)

            if token_id in EOS_TOKEN_IDS:
                break

        elapsed = time.time() - start
        print(f"\n  [{token_count} tokens, {elapsed:.1f}s, {token_count/elapsed:.1f} tok/s]")

        assistant_text = tokenizer.decode(generated_ids, skip_special_tokens=True)
        messages.append({"role": "assistant", "content": assistant_text})

    model.free()


def main():
    parser = argparse.ArgumentParser(description="Interactive Qwen3-0.6B chat")
    parser.add_argument("--max-tokens", type=int, default=512, help="Max tokens per response")
    parser.add_argument("--no-think", action="store_true", help="Disable thinking mode")
    parser.add_argument("--gpu", type=int, default=None, help="GPU device ID (default: GLM_GPU env var or 0)")
    parser.add_argument("--max-seq-len", type=int, default=2048, help="Max sequence length")
    args = parser.parse_args()

    gpu_id = args.gpu if args.gpu is not None else int(os.environ.get("GLM_GPU", "0"))
    os.environ["CUDA_VISIBLE_DEVICES"] = str(gpu_id)

    print(f"Loading model on GPU {gpu_id}...")
    glm = GlmOps(device_id=0)
    model = Qwen3Model.from_pretrained(glm, QWEN3_REPO, max_batch=1, max_seq_len=args.max_seq_len)
    tokenizer = AutoTokenizer.from_pretrained(QWEN3_REPO)

    print(f"Qwen3-0.6B ready (max_tokens={args.max_tokens}, thinking={'off' if args.no_think else 'on'})")
    print("Type a message to chat. /clear to reset, /q to quit.")

    chat_loop(model, tokenizer, args.max_tokens, args.no_think)


if __name__ == "__main__":
    main()

File: src/run_qwen3.ts
import { GlmOps } from "./glm_ops";
import { Qwen3Model } from "./qwen3_model";
import { FlatKVCache } from "./flat_kv";
import { AutoTokenizer, PreTrainedTokenizer } from "@huggingface/transformers";
import { resolveModelPath } from "./model_path";
import { createInterface } from "node:readline";

const QWEN3_REPO = "Qwen/Qwen3-0.6B";
const QWEN3_FP8_REPO = "Qwen/Qwen3-0.6B-FP8";
const EOS_TOKEN_IDS = new Set([151645, 151643]);

async function chatLoop(model: Qwen3Model, cache: FlatKVCache, tokenizer: PreTrainedTokenizer, maxNewTokens: number, noThink: boolean): Promise<void> {
  const messages: Array<{ role: string; content: string }> = [];
  const enableThinking = !noThink;
  const rl = createInterface({ input: process.stdin, output: process.stdout });

  const askQuestion = (prompt: string): Promise<string> =>
    new Promise((resolve) => rl.question(prompt, resolve));

  while (true) {
    const userInput = (await askQuestion("\nYou: ")).trim();
    if (!userInput) continue;
    if (userInput.toLowerCase() === "quit" || userInput.toLowerCase() === "exit" || userInput.toLowerCase() === "/q") break;
    if (userInput.toLowerCase() === "/clear") {
      messages.length = 0;
      console.log("Conversation cleared.");
      cache.reset();
      continue;
    }

    messages.push({ role: "user", content: userInput });

    const result = tokenizer.apply_chat_template(messages, {
      tokenize: true,
      add_generation_prompt: true,
      return_tensor: false,
      return_dict: true,
      tokenizer_kwargs: { enable_thinking: enableThinking },
    }) as { input_ids: number[] | number[][] };
    const inputIds = (Array.isArray(result.input_ids[0]) ? result.input_ids : [result.input_ids]) as number[][];

    if (inputIds[0].length > model.maxSeqLen) {
      console.log(`Prompt (${inputIds[0].length} tokens) exceeds max_seq_len (${model.maxSeqLen}). Truncating conversation.`);
      messages.pop();
      continue;
    }

    process.stdout.write("\nAssistant: ");
    const generatedIds: number[] = [];
    const start = Date.now();
    let tokenCount = 0;

    for (const tokenId of model.streamTokens(inputIds, cache, maxNewTokens, EOS_TOKEN_IDS)) {
      generatedIds.push(tokenId);
      tokenCount++;
      const chunk = tokenizer.decode([tokenId], { skip_special_tokens: false });
      process.stdout.write(chunk);
      if (EOS_TOKEN_IDS.has(tokenId)) break;
    }

    const elapsed = (Date.now() - start) / 1000;
    console.log(`\n  [${tokenCount} tokens, ${elapsed.toFixed(1)}s, ${(tokenCount / elapsed).toFixed(1)} tok/s]`);

    const assistantText = tokenizer.decode(generatedIds, { skip_special_tokens: true });
    messages.push({ role: "assistant", content: assistantText });
  }

  rl.close();
  model.free();
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  let maxTokens = 512;
  let noThink = false;
  let gpu: number | undefined;
  let maxSeqLen = 2048;
  let useFp8 = false;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--max-tokens" && i + 1 < args.length) {
      maxTokens = parseInt(args[++i], 10);
    } else if (args[i] === "--no-think") {
      noThink = true;
    } else if (args[i] === "--gpu" && i + 1 < args.length) {
      gpu = parseInt(args[++i], 10);
    } else if (args[i] === "--max-seq-len" && i + 1 < args.length) {
      maxSeqLen = parseInt(args[++i], 10);
    } else if (args[i] === "--fp8") {
      useFp8 = true;
    }
  }

  const gpuId = gpu ?? parseInt(process.env.GLM_GPU ?? "0", 10);
  process.env.CUDA_VISIBLE_DEVICES = String(gpuId);

  const repoId = useFp8 ? QWEN3_FP8_REPO : QWEN3_REPO;

  console.log(`Loading model on GPU ${gpuId}${useFp8 ? " (FP8)" : ""}...`);
  const glm = new GlmOps(0);
  const model = Qwen3Model.fromPretrained(glm, repoId, 1, maxSeqLen);
  const cache = model.createFlatKVCache();

  const modelDir = resolveModelPath(repoId);
  const tokenizer = await AutoTokenizer.from_pretrained(modelDir, { local_files_only: true });

  console.log(`Qwen3-0.6B ready (max_tokens=${maxTokens}, thinking=${noThink ? "off" : "on"}${useFp8 ? ", fp8" : ""})`);
  console.log("Type a message to chat. /clear to reset, /q to quit.");

  await chatLoop(model, cache, tokenizer, maxTokens, noThink);
}

main().catch((err) => {
  console.error("Failed:", err);
  process.exit(1);
});
